### PR TITLE
Add metric to track the size and labels of individual disks

### PR DESCRIPTION
**Pull request by paulajulve**
Re: https://github.com/grafana/deployment_tools/issues/111858

Once the disks are detached from the cluster, the only way back to knowing where they belonged is by using the namespace and the zone. But some namespaces exist in different zones/clusters (ie, generic names like `hosted-exporters`). The current `unused_disks_size_gb` metric was adding up the sizes of all disks under the same namespace. In the case of unique namespaces, that's alright. But in the case of those that span multiple regions, it can lead to costs being associated to the wrong cluster, and somewhat difficult to interpret data.

To make it easier for future us, I'm adding a new `unused_disk_size_bytes` metric here, that publishes the size data per disk, instead of added up. It still carries the `k8s_namespace`, `type`, and `zone` labels, since we'll need all of them to be able to attribute costs appropriately.

I've also opted for bytes as the unit, since the size field is not uniform across providers. Azure and GCP offer the size of the volume in GB, while AWS uses GiB. Even though they all end up billing per GB 🙄 

I've tested this out locally both for AWS and GCP. I have not tested it for Azure. I'll work on that now.

### CODE REVIEW

**Comment by paulajulve** [comment]
Oh, I've just realised this one will require updating the readme also. I'll work on that.

In the meantime, @inkel , I'm slightly concerned about this warning [here](https://github.com/grafana/unused/blob/main/README.md?plain=1#L58). Does this still hold true? I don't fully understand what it means.

**Comment by inkel** [comment]
> Oh, I've just realised this one will require updating the readme also. I'll work on that.
> 
> In the meantime, @inkel , I'm slightly concerned about this warning [here](https://github.com/grafana/unused/blob/main/README.md?plain=1&rgh-link-date=2024-04-08T12%3A59%3A06Z#L58). Does this still hold true? I don't fully understand what it means.

@paulajulve I believe it still does. That particular comment refers to adding the disk ID as a label as the main culprit to leading to cardinality explosion of metrics. Say for instance you have 1,000 unused disks on each provider, this means you'll end up with a label with _1,000 times the number of provider_ unique values. As per [this article](https://www.robustperception.io/cardinality-is-key/) label cardinality should be as low as possible if we don't want to hog down Prometheus.

File: aws/disk.go
Comment (inkel) [reply aws/disk.go:51]: Agreed with @jjo. Also note that here you're doing float64 to int just to do int to float64 when you emit the metric, so that's two unnecessary conversions.

File: gcp/disk.go
Comment (Pokom) [reply gcp/disk.go:56]: Nit: Providers charge per GiB, which is `1024 ^3`

File: unusedtest/disk.go
Comment (Pokom) [reply unusedtest/disk.go:34]: Ditto nit: Providers charge per GiB, which is 1024 ^3

File: gcp/disk.go
Comment (inkel) [reply gcp/disk.go:56]: In addition to being simpler to read, using `math.Pow` is significantly less performant:

```
inkel@numenor:~/go/src/github.com/inkel/benchmarks$ go test -bench=BenchmarkPow -benchmem
goos: darwin
goarch: arm64
pkg: github.com/inkel/benchmarks
BenchmarkPow/1000*1000*1000-10          1000000000               0.3217 ns/op          0 B/op          0 allocs/op
BenchmarkPow/math.Pow-10                125777395                9.544 ns/op           0 B/op          0 allocs/op
PASS
ok      github.com/inkel/benchmarks     2.779s
```

**Review by inkel — COMMENTED** [review]
As explained on the main comments section, I'm concerned about the cardinality of the new metric, in particular in the `disk` label.

Given that the information was logged we could use Loki and its [`unwrap` LogQL keyword](https://grafana.com/docs/loki/latest/query/query_examples/#unwrap-examples) to create values from these, instead of loading Prometheus.

**Comment by paulajulve** [comment]
I could update the "unused disk found" log to include the k8s_namespace too (and maybe use bytes instead of GB/GiB) and use unwrap more or less like this ([explore](https://ops.grafana-ops.net/explore?schemaVersion=1&panes=%7B%22btb%22:%7B%22datasource%22:%22c-R8UWvVk%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bcluster%3D%5C%22prod-us-east-0%5C%22,namespace%3D%5C%22unused-exporter%5C%22,%20container%21%3D%5C%22eventrouter%5C%22%7D%20%7C%20logfmt%20%7C%20msg%3D%5C%22unused%20disk%20found%5C%22%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22c-R8UWvVk%22%7D,%22editorMode%22:%22code%22,%22hide%22:false%7D,%7B%22refId%22:%22B%22,%22expr%22:%22sum%20by%20%28k8s_namespace,%20zone%29%20%28%5Cn%20%20sum_over_time%28%5Cn%20%20%20%20%7Bnamespace%3D%5C%22unused-exporter%5C%22,%20container%21%3D%5C%22eventrouter%5C%22%7D%5Cn%20%20%20%20%20%20%7C%20logfmt%5Cn%20%20%20%20%20%20%7C%20unwrap%20size_bytes%20%5B1m%5D%29%5Cn%20%20%29%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22c-R8UWvVk%22%7D,%22editorMode%22:%22code%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D%7D&orgId=1)):
```
sum by (k8s_namespace, zone) (
  sum_over_time(
    {namespace="unused-exporter", container!="eventrouter"}
      | logfmt
      | unwrap size_bytes [1m])
  )
```

I'm unclear on next steps tho 🤔 we want to add this to our TCO recording rules. Can we add this expression to a prometheus recording rule? I've seen loki based alerts, but I'm not sure we can mix and match like this out of the box.

**Comment by jjo** [comment]
> 
> I'm unclear on next steps tho 🤔 we want to add this to our TCO recording rules. Can we add this expression to a prometheus recording rule? I've seen loki based alerts, but I'm not sure we can mix and match like this out of the box.

As these are metrics we're adding, it'll be a matter of juggling how we could come up with a recording rule to "fill the gap(s)" of the missing labels to join, `cluster` in particular as a very relevant one.

Crafting a promQL to join the metrics we're adding to some existing ones (note there I _mocked_ 3 existing disks ab-using `absent(...)`, as if those metrics were generated by unused exporter after merging this PR):

[/explore](https://ops.grafana-ops.net/explore?schemaVersion=1&panes=%7B%224p8%22:%7B%22datasource%22:%22000000134%22,%22queries%22:%5B%7B%22refId%22:%22B%22,%22expr%22:%22%23%20Ultimate%20goal:%20account%20%28aggregate%29%20disk%20GB%20by:%20cluster,namespace,team%5Cn%23%5Cn%23%20NOTE:%20all%20label_join%28%29%20calls%20are%20there%20to%20%5C%22normalize%5C%22%20existing%20labels%20towards:%20cluster,%20namespace,%20zone%5Cn%23%5Cnsum%20by%20%28cluster,%20namespace,%20team%29%28%5Cn%20%20%20%20%23%20Mock%203%20disks%20on%20respective%20ns,zone%20with%2010GB,%2010GB,%20100GB%5Cn%20%20%20%20label_join%28%20%20%23%20k8s_namespace%20-%3E%20namespace%5Cn%20%20%20%20%20%20%20%20sum%20by%20%28k8s_namespace,%20zone%29%28%5Cn%20%20%20%20%20%20%20%20%20%20%20%2010e9%20%2A%20absent%28%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20unused_disk_size_bytes%7Bk8s_namespace%3D%5C%22loki-ops%5C%22,%20type%3D%5C%22ssd%5C%22,%20zone%3D%5C%22us-east4%5C%22,%20id%3D%5C%22what-ever-001%5C%22%7D%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%29%20or%5Cn%20%20%20%20%20%20%20%20%20%20%20%2010e9%20%2A%20absent%28%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20unused_disk_size_bytes%7Bk8s_namespace%3D%5C%22loki-ops%5C%22,%20type%3D%5C%22ssd%5C%22,%20zone%3D%5C%22us-east4%5C%22,%20id%3D%5C%22what-ever-002%5C%22%7D%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%29%20or%5Cn%20%20%20%20%20%20%20%20%20%20%20%20100e9%20%2A%20absent%28%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20unused_disk_size_bytes%7Bk8s_namespace%3D%5C%22cortex-prod-10%5C%22,%20type%3D%5C%22ssd%5C%22,%20zone%3D%5C%22us-central1%5C%22,%20id%3D%5C%22what-ever-002%5C%22%7D%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%29%5Cn%20%20%20%20%20%20%20%20%29,%20%5C%22namespace%5C%22,%20%5C%22%5C%22,%20%5C%22k8s_namespace%5C%22%5Cn%20%20%20%20%29%5Cn%5Cn%20%20%20%20%23%20Join%20the%20%28few%29%20labels%20we%20have%20from%20%60unused_disk_size_bytes%60%20metric%20against%20those%20from%20tanka%5Cn%20%20%20%20%2A%20on%20%28namespace,%20zone%29%20group_right%28%29%5Cn%5Cn%20%20%20%20label_join%28%20%20%20%20%20%20%20%20%20%23%20location%20-%3E%20zone%5Cn%20%20%20%20%20%20%20%20label_join%28%20%20%20%20%20%23%20exported_cluster%20-%3E%20cluster%5Cn%20%20%20%20%20%20%20%20%20%20%20%20label_join%28%20%23%20exported_namespace%20-%3E%20namespace%5Cn%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20HACK:%20%28just%29%20pick%20_one_%20cluster%20from%20the%20zone,%20as%20%60unused_disk_size_bytes%60%20doesn%27t%20have%20%5C%22cluster%5C%22%20label%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20max%20by%20%28exported_cluster,%20exported_namespace,%20location,%20team%29%28%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20tanka_environment_info%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2A%20on%20%28exported_cluster%29%20group_left%28location%29%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20tanka_cluster_info%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%29,%20%5C%22namespace%5C%22,%20%5C%22%5C%22,%20%5C%22exported_namespace%5C%22%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%29,%20%5C%22cluster%5C%22,%20%5C%22%5C%22,%20%5C%22exported_cluster%5C%22%5Cn%20%20%20%20%20%20%20%20%29,%20%5C%22zone%5C%22,%20%5C%22%5C%22,%20%5C%22location%5C%22%5Cn%20%20%20%20%29%5Cn%29%20%2F%201e9%20%23%20show%20the%20sum%20in%20GB%22,%22range%22:false,%22instant%22:true,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22000000134%22%7D,%22editorMode%22:%22code%22,%22legendFormat%22:%22__auto%22,%22exemplar%22:false%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D%7D&orgId=1)
```
# Ultimate goal: account (aggregate) disk GB by: cluster,namespace,team
#
# NOTE: all label_join() calls are there to "normalize" existing labels towards: cluster, namespace, zone
#
sum by (cluster, namespace, team)(
    # Mock 3 disks on respective ns,zone with 10GB, 10GB, 100GB
    label_join(  # k8s_namespace -> namespace
        sum by (k8s_namespace, zone)(
            10e9 * absent(
                unused_disk_size_bytes{k8s_namespace="loki-ops", type="ssd", zone="us-east4", id="what-ever-001"}
            ) or
            10e9 * absent(
                unused_disk_size_bytes{k8s_namespace="loki-ops", type="ssd", zone="us-east4", id="what-ever-002"}
            ) or
            100e9 * absent(
                unused_disk_size_bytes{k8s_namespace="cortex-prod-10", type="ssd", zone="us-central1", id="what-ever-002"}
            )
        ), "namespace", "", "k8s_namespace"
    )

    # Join the (few) labels we have from `unused_disk_size_bytes` metric against those from tanka
    * on (namespace, zone) group_right()

    label_join(         # location -> zone
        label_join(     # exported_cluster -> cluster
            label_join( # exported_namespace -> namespace

                # HACK: (just) pick _one_ cluster from the zone, as `unused_disk_size_bytes` doesn't have "cluster" label
                max by (exported_cluster, exported_namespace, location, team)(
                    tanka_environment_info
                    * on (exported_cluster) group_left(location)
                    tanka_cluster_info
                ), "namespace", "", "exported_namespace"
            ), "cluster", "", "exported_cluster"
        ), "zone", "", "location"
    )
) / 1e9 # show the sum in GB
```
ends up showing:

image

Then, with the `(cluster, namespace, team)` grouping we should be ready to aggregate them into a cost recording-rule, as decided in [Design Doc: include unused disks on TCO d/doc](https://docs.google.com/document/d/1teLO7_jYbl9T89ajMD2hlFWkNhSOEnvrqF70VQ97ulc/edit#heading=h.5sybau7waq2q)

**Comment by jjo** [comment]
> As explained on the main comments section, I'm concerned about the cardinality of the new metric, in particular in the `disk` label.
> 
> Given that the information was logged we could use Loki and its [`unwrap` LogQL keyword](https://grafana.com/docs/loki/latest/query/query_examples/#unwrap-examples) to create values from these, instead of loading Prometheus.

IMO these disk IDs live in the ~same playground as pod IDs (which of course we already carry from KSM metrics), and I think we may need them as _metrics_, ready there to be able to create tables (showing `ID`, `size`) for teams to prioritize their cleanup.
Regarding cardinality, I'm not much worried about it in this use case, because there's no `O(n²)` (or worse) as there's 1:1 `ID<->disk` (plus other labels) without further permutations, and we just have a thousand ~ish of these.

File: aws/disk.go
Comment (paulajulve) [reply aws/disk.go:51]: > you're doing float64 to int just to do int to float64

No, I'm not. The Size I get straight from the object here is an int32. Which is too small to fit the values in bytes. Granted, I could cast it straight to float64 and skip one cast. I was trying to keep consistency to the other size method. But I can change it to float64.

File: gcp/disk.go
Comment (paulajulve) [reply gcp/disk.go:56]: > Nit: Providers charge per GiB, which is 1024 ^3

Where did you find that? 👀  I tried to check and all the pricing I found was in GB. Even the AWS one, even though they provide the size of the volumes on GiB.

File: unusedtest/disk.go
Comment (paulajulve) [reply unusedtest/disk.go:34]: I'm going for bytes so we guarantee consistency on the metrics we get, since AWS provides the volume sizes in GiB, while GCP uses GB 🙄 Surprisingly enough, Azure is the only one that provides both values out of the box.

File: gcp/disk.go
Comment (Pokom) [reply gcp/disk.go:56]: They don't make it easy and I only know this because I've spent way too much time reading the pricing pages for each CSP. 

---

Google: https://cloud.google.com/compute/disks-image-pricing#disk

> For Compute Engine, disk size, machine type memory, and network usage are calculated in JEDEC binary gigabytes (GB), or IEC gibibytes (GiB), where 1 GiB is 230 bytes. Similarly, 1 TiB is 240 bytes, or 1024 JEDEC GBs.

---

AWS: https://aws.amazon.com/s3/pricing/

---

Azure: https://azure.microsoft.com/en-us/pricing/details/storage/blobs/#pricing

> 

Azure Data Lake Storage usage is calculated in binary Gigabytes (GB), where 1 GB = 230 bytes. This unit of measurement is also known as Gibibyte (GiB), defined by the International Electrotechnical Commission (IEC). Similarly, 1 TB is 240 bytes, i.e. 1,024 GBs.




> Amazon S3 storage usage is calculated in binary gigabytes (GB), where 1 GB is 230 bytes. This unit of measurement is also known as a gibibyte (GiB), defined by the International Electrotechnical Commission (IEC). Similarly, 1 TB is 240 bytes, i.e. 1024 GBs.

Comment (inkel) [reply gcp/disk.go:56]: Am I the only one finding AWS and Azure text TOO similar, as in one copied the other and only changed their product name?

File: aws/disk.go
Comment (jjo) [inline aws/disk.go:50]: Sorry for the "2nd" one on this: maybe add 
```
constant (
  GiBbytes = 1073741824
  GBbytes = 1000000000
)
```
at the top, and use them?

File: cmd/unused-exporter/exporter.go
Comment (jjo) [inline cmd/unused-exporter/exporter.go:288]: ❤️

File: cmd/unused-exporter/exporter.go
Comment (paulajulve) [reply cmd/unused-exporter/exporter.go:70]: All metrics carry the provider label by default (defined [here](https://github.com/grafana/unused/blob/main/cmd/unused-exporter/exporter.go#L45)). This line only adds the extra labels. _However_, this provider label is not using that convention. The values it currently assigns are: `aws`, `gcp`, and `azure`. Either we overwrite that, breaking consistency with the rest of the project (spoiler, I don't like this solution, especially given that this is a public repo), or we gotta do some fun with label replacing on those queries.

Comment (jjo) [reply cmd/unused-exporter/exporter.go:70]: oh cool, and +1 not changing them / their default: I'll create a PR to make them configurable via flags, as it makes sense for them to be "fixed" depending on the env you're running it, and other CSP/provider naming you'd be using.

**Review by jjo — APPROVED** [review]
LGTM semantically and syntactically up to my (GOlang) read/ability, but let's better also get @inkel 's approval for the latter.

**Review by inkel — APPROVED** [review]
LGTM.

File: cmd/unused-exporter/exporter.go
Comment (jjo) [reply cmd/unused-exporter/exporter.go:70]: The PR https://github.com/grafana/unused/pull/92
/cc @inkel

**Pull request merged**
